core/src/main/java/org/infinispan/stream/impl/LockStreamImpl.java

```diff
@@ -0,0 +1,231 @@
+   public BaseCacheStream distributedBatchSize(int batchSize) {
+      return newOrReuse(realStream.distributedBatchSize(batchSize));
+   }
+
+   @Override
+   public BaseCacheStream segmentCompletionListener(SegmentCompletionListener listener) {
+      throw new UnsupportedOperationException("LockStream doesn't support completion listener");
+   }
+
+   @Override
+   public BaseCacheStream disableRehashAware() {
+      return newOrReuse(realStream.disableRehashAware());
+   }
+
+   @Override
+   public BaseCacheStream timeout(long timeout, TimeUnit unit) {
+      return newOrReuse(realStream.timeout(timeout, unit));
+   }
+
+   @Override
+   public Iterator<CacheEntry<K, V>> iterator() {
+      return realStream.iterator();
+   }
+
+   @Override
+   public Spliterator<CacheEntry<K, V>> spliterator() {
+      return realStream.spliterator();
+   }
+
+   @Override
+   public boolean isParallel() {
+      return realStream.isParallel();
+   }
+
+   @Override
+   public LockedStream<K, V> sequential() {
+      return newOrReuse(realStream.sequential());
+   }
+
+   @Override
+   public LockedStream<K, V> parallel() {
+      return newOrReuse(realStream.parallel());
+   }
+
+   @Override
+   public LockedStream<K, V> unordered() {
+      // This stream is always unordered
+      return this;
+   }
+
+   @Override
+   public LockedStream<K, V> onClose(Runnable closeHandler) {
+      return newOrReuse(realStream.onClose(closeHandler));
+   }
+
+   @Override
+   public void close() {
+      realStream.close();
+   }
+
+   @SerializeWith(value = CacheEntryConsumer.CacheEntryConsumerExternalizer.class)
+   private static class CacheEntryConsumer<K, V> implements BiConsumer<Cache<K, V>, CacheEntry<K, V>>, Serializable {
+      private final BiConsumer<Cache<K, V>, ? super CacheEntry<K, V>> realConsumer;
+      private final Predicate<? super CacheEntry<K, V>> predicate;
+      private transient LockManager lockManager;
+
+      private CacheEntryConsumer(BiConsumer<Cache<K, V>, ? super CacheEntry<K, V>> realConsumer,
+            Predicate<? super CacheEntry<K, V>> predicate) {
+         this.realConsumer = realConsumer;
+         this.predicate = predicate;
+      }
+
+      @Override
+      public void accept(Cache<K, V> kvCache, CacheEntry<K, V> kvCacheEntry) {
+         K key = kvCacheEntry.getKey();
+         lock(key);
+         try {
+            CacheEntry<K, V> rereadEntry = kvCache.getAdvancedCache().getCacheEntry(key);
+            if (rereadEntry != null && (predicate == null || predicate.test(rereadEntry))) {
+               // Pass the Cache with the owner set to our key so they can write and also it won't unlock that key
+               realConsumer.accept(kvCache.getAdvancedCache().lockAs(key), kvCacheEntry);
+            }
+         } finally {
+            lockManager.unlock(key, key);
+         }
+      }
+
+      private void lock(K key) {
+         KeyAwareLockPromise kalp = lockManager.lock(key, key, 10, TimeUnit.SECONDS);
```

Comment:
The only issue I originally had is because the timeout is only _used_ with transactional caches. So setting this on a non tx cache for this purpose seemed a bit weird. But I can change it to use it either way.